src/runtime/pkg/kata-monitor/cri.go

```diff
@@ -0,0 +1,200 @@
+	if protocol, addr, err = parseEndpoint(endpoint); err != nil && protocol == "" {
+		fallbackEndpoint := fallbackProtocol + "://" + endpoint
+		protocol, addr, err = parseEndpoint(fallbackEndpoint)
+		if err == nil {
+			monitorLog.Warningf("Using %q as endpoint is deprecated, please consider using full url format %q.", endpoint, fallbackEndpoint)
+		}
+	}
+	return
+}
+
+func parseEndpoint(endpoint string) (string, string, error) {
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return "", "", err
+	}
+
+	switch u.Scheme {
+	case "tcp":
+		return "tcp", u.Host, nil
+
+	case "unix":
+		return "unix", u.Path, nil
+
+	case "":
+		return "", "", fmt.Errorf("using %q as endpoint is deprecated, please consider using full url format", endpoint)
+
+	default:
+		return u.Scheme, "", fmt.Errorf("protocol %q not supported", u.Scheme)
+	}
+}
+
+// getSandboxes get kata sandbox from the container engine.
+// this will be called only after monitor start.
+func (km *KataMonitor) getSandboxes() (map[string]struct{}, error) {
+
+	sandboxMap := make(map[string]struct{})
+	runtimeClient, runtimeConn, err := getRuntimeClient(km.runtimeEndpoint)
+	if err != nil {
+		return sandboxMap, err
+	}
+	defer closeConnection(runtimeConn)
+
+	filter := &pb.PodSandboxFilter{
+		State: &pb.PodSandboxStateValue{
+			State: pb.PodSandboxState_SANDBOX_READY,
+		},
+	}
+
+	request := &pb.ListPodSandboxRequest{
+		Filter: filter,
+	}
+	monitorLog.Debugf("ListPodSandboxRequest: %v", request)
+	r, err := runtimeClient.ListPodSandbox(context.Background(), request)
+	if err != nil {
+		return sandboxMap, err
+	}
+	monitorLog.Debugf("ListPodSandboxResponse: %v", r)
+
+	for _, pod := range r.Items {
+		request := &pb.PodSandboxStatusRequest{
+			PodSandboxId: pod.Id,
+			Verbose:      true,
+		}
+
+		r, err := runtimeClient.PodSandboxStatus(context.Background(), request)
+		if err != nil {
+			return sandboxMap, err
+		}
+
+		lowRuntime := ""
+		var res map[string]interface{}
+		if err := json.Unmarshal([]byte(r.Info["info"]), &res); err != nil {
+			monitorLog.WithError(err).WithField("pod", r).Error("failed to Unmarshal pod info")
+			continue
+		} else {
+			monitorLog.WithField("pod info", res).Debug("")
+
+			// get low level container runtime
+			// containerd stores the pod runtime in "/runtimeType" while CRI-O stores it the
+			// io.kubernetes.cri-o.RuntimeHandler annotation: check for both.
+			keys := []string{"/runtimeType", "/runtimeSpec/annotations/io.kubernetes.cri-o.RuntimeHandler"}
+			for _, key := range keys {
+				pointer, _ := gojsonpointer.NewJsonPointer(key)
+				rt, _, _ := pointer.Get(res)
+				if rt != nil {
+					if str, ok := rt.(string); ok {
+						lowRuntime = str
+						break
+					}
+				}
+			}
+		}
+
+		// If lowRuntime is empty something changed in containerd/CRI-O or we are dealing with an unknown container engine.
+		// Safest options is to add the POD in the list: we will be able to connect to the shim to retrieve the actual info
```

Comment:
What happens when we try to dial a runtime that doesn't exist? Will we _then_ drop it from the list?

I agree with Marcel -- info will be lost in the noise.  This is an unexpected event or a misconfiguration. We should have this as a warning.

Reply:
Is this the same logic that was applied in the containers-only case, too?